internal/webhook/external/azure.go

```diff
@@ -0,0 +1,219 @@
+}
+
+// getHandler implements WebhookReceiver.
+func (a *azureWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("received azure event", "userAgent", r.UserAgent())
+
+		switch {
+		// format is AzureContainerRegistry/<version>
+		case strings.Contains(r.UserAgent(), "AzureContainerRegistry"):
+			a.handleAcrEvent(ctx, w, requestBody)
+		// Format is VSServices/<version>
+		case strings.Contains(r.UserAgent(), "VSServices"):
+			a.handleAzureDevOpsEvent(ctx, w, requestBody)
+		default:
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					errors.New(
+						"request does not appear to have originated from a supported service",
+					),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+	})
+}
+
+func (a *azureWebhookReceiver) handleAcrEvent(
+	ctx context.Context,
+	w http.ResponseWriter,
+	requestBody []byte,
+) {
+	var event acrEvent
+	if err := json.Unmarshal(requestBody, &event); err != nil {
+		xhttp.WriteErrorJSON(
+			w,
+			xhttp.Error(errors.New("invalid request body"),
+				http.StatusBadRequest,
+			),
+		)
+		return
+	}
+
+	switch event.Action {
+	case "push":
+		repoURLs := resolveAcrRepoURLs(event)
+		logger := logging.LoggerFromContext(ctx)
+		logger = logger.WithValues("repoURLs", repoURLs)
+		ctx = logging.ContextWithLogger(ctx, logger)
+		refreshWarehouses(ctx, w, a.client, a.project, repoURLs...)
+	case "ping":
+		xhttp.WriteResponseJSON(
+			w,
+			http.StatusOK,
+			map[string]string{
+				"msg": "ping event received, webhook is configured correctly",
+			},
+		)
+	default:
+		xhttp.WriteErrorJSON(
+			w,
+			xhttp.Error(
+				fmt.Errorf("event type %s is not supported", event.Action),
+				http.StatusBadRequest,
+			),
+		)
+		return
+	}
+}
+
+func (a *azureWebhookReceiver) handleAzureDevOpsEvent(
+	ctx context.Context,
+	w http.ResponseWriter,
+	requestBody []byte,
+) {
+	var event azureDevOpsEvent
+	if err := json.Unmarshal(requestBody, &event); err != nil {
+		xhttp.WriteErrorJSON(
+			w,
+			xhttp.Error(errors.New("invalid request body"),
+				http.StatusBadRequest,
+			),
+		)
+		return
+	}
+	repoURL := git.NormalizeURL(event.Resource.Repository.RemoteURL)
+	logger := logging.LoggerFromContext(ctx)
+	logger = logger.WithValues("repoURL", repoURL)
+	ctx = logging.ContextWithLogger(ctx, logger)
+	refreshWarehouses(ctx, w, a.client, a.project, repoURL)
+}
+
+func resolveAcrRepoURLs(e acrEvent) []string {
+	repoURL := fmt.Sprintf("%s/%s", e.Request.Host, e.Target.Repository)
+	var repoURLs []string
+	switch e.Target.MediaType {
+	case imageMediaType:
+		repoURLs = append(repoURLs, image.NormalizeURL(repoURL))
+	case helmChartMediaType:
+		repoURLs = append(repoURLs, helm.NormalizeChartRepositoryURL(repoURL))
+	default:
+		repoURLs = append(repoURLs,
+			image.NormalizeURL(repoURL),
+			helm.NormalizeChartRepositoryURL(repoURL),
+		)
+	}
```

Comment:
If we don't recognize the media type, I think we actually have _nothing_ to do.

I'd suggest changing the return type of this function from `[]string` to `string`, since after this change, we'd return one URL _at most_.

Reply:
Eh... when the return type changes, don't forget to de-pluralize the function name.

Reply:
Ok... if we can get #4621 merged first, you can rebase this PR on main or merge the latest from main into it and you'll have access to an improved way of normalizing OCI repo URLs that takes all relevant media types into account.